### PR TITLE
chore(asm): force compliance and improve regex for query strings

### PR DESCRIPTION
**Pull request by christophe-papazian**
RFC for **QueryString - Sensitive Data Obfuscation** is specifying to use `DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP` for query strings obfuscation.

1. deprecation of `DD_TRACE_OBFUSCATION_QUERY_STRING_PATTERN`. Use `DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP` instead.
2. Updated the default regex pattern to use the newer, improved version of the pattern, as specified in the RFC. In particular:
    1. handles trailing dashes and newline for private keys
    2. handles DSA and ECDSA SSH public keys
    3. handles base64 trailing equals for SSH public keys
    4. handles SSH public key comments
    5. handles typical password change keys
    6. catches the leading quote for JSON-ish values 
4. Updated tests and docs 
5. Use new system tests to reflect the changes [PR 1537](https://github.com/DataDog/system-tests/pull/1537)

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

**Comment by christophe-papazian** [comment]
> We need to deprecate the existing environment variable first since this would be a breaking change.

@majorgreys done.

**Review by ZStriker19 — COMMENTED** [review]
Now that we've changed the default, should we add some tests that verify the behavior of the new regex? Or is that covered by system-tests or something else?

File: ddtrace/settings/config.py [hunk 1]
```diff
@@ -334,9 +366,18 @@ def __init__(self):
...
+        if "DD_TRACE_OBFUSCATION_QUERY_STRING_PATTERN" in os.environ:
+            deprecate(
+                "`DD_TRACE_OBFUSCATION_QUERY_STRING_PATTERN` is deprecated "
+                "and will be removed in the next major version.",
+                message="use `DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP` instead",
+                removal_version="2.0.0",
+            )
+            dd_trace_obfuscation_query_string_pattern = os.getenv("DD_TRACE_OBFUSCATION_QUERY_STRING_PATTERN")
+        else:
+            dd_trace_obfuscation_query_string_pattern = os.getenv(
```
Comment (majorgreys) [inline ddtrace/settings/config.py:378]: Non-blocking: Rename this internal attribute to the new name

File: docs/configuration.rst [hunk 2]
```diff
@@ -286,10 +286,14 @@ The following environment variables for the tracer are supported:
...
+         ``'(?ix)(?:(?:"|%22)?)(?:(?:old[-_]?|new[-_]?)?p(?:ass)?w(?:or)?d(?:1|2)?|pass(?:[-_]?phrase)?|secret|(?:api[-_]?|private[-_]?|public[-_]?|access[-_]?|secret[-_]?)key(?:[-_]?id)?|token|consumer[-_]?(?:id|key|secret)|sign(?:ed|ature)?|auth(?:entication|orization)?)(?:(?:\\s|%20)*(?:=|%3D)[^&]+|(?:"|%22)(?:\\s|%20)*(?::|%3A)(?:\\s|%20)*(?:"|%22)(?:%2[^2]|%[^2]|[^"%])+(?:"|%22))|(?: bearer(?:\\s|%20)+[a-z0-9._\\-]+|token(?::|%3A)[a-z0-9]{13}|gh[opsu]_[0-9a-zA-Z]{36}|ey[I-L](?:[\\w=-]|%3D)+\\.ey[I-L](?:[\\w=-]|%3D)+(?:\\.(?:[\\w.+/=-]|%3D|%2F|%2B)+)?|-{5}BEGIN(?:[a-z\\s]|%20)+PRIVATE(?:\\s|%20)KEY-{5}[^\\-]+-{5}END(?:[a-z\\s]|%20)+PRIVATE(?:\\s|%20)KEY(?:-{5})?(?:\\n|%0A)?|(?:ssh-(?:rsa|dss)|ecdsa-[a-z0-9]+-[a-z0-9]+)(?:\\s|%20|%09)+(?:[a-z0-9/.+]|%2F|%5C|%2B){100,}(?:=|%3D)*(?:(?:\\s|%20|%09)+[a-z0-9._-]+)?)'``
      description: A regexp to redact sensitive query strings. Obfuscation disabled if set to empty string
+     version_added:
+       v1.19.0: |
+           DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP replace DD_TRACE_OBFUSCATION_QUERY_STRING_PATTERN that becomes deprecated
+           and will be deleted in 2.0.0
```
Comment (majorgreys) [inline docs/configuration.rst:296]:
```suggestion
           ``DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP`` replaces ``DD_TRACE_OBFUSCATION_QUERY_STRING_PATTERN`` which is deprecated
           and will be deleted in 2.0.0
```

File: releasenotes/notes/querystring_improved_sensitive_data_information-b055d0cda09da1c5.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+other:
+  - |
+    ASM: Improve default value of regex for query string obfuscation.
+    Rename env var `DD_TRACE_OBFUSCATION_QUERY_STRING_PATTERN` to `DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP`.
```
Comment (majorgreys) [inline releasenotes/notes/querystring_improved_sensitive_data_information-b055d0cda09da1c5.yaml:5]:
```suggestion
    Rename env var ``DD_TRACE_OBFUSCATION_QUERY_STRING_PATTERN`` to ``DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP``.
```

File: ddtrace/settings/config.py (hunk 1 above)
Comment (christophe-papazian) [reply ddtrace/settings/config.py]: done

File: ddtrace/settings/config.py [hunk 4]
```diff
@@ -334,20 +366,29 @@ def __init__(self):
...
+                "DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP", DD_TRACE_OBFUSCATION_QUERY_STRING_REGEXP_DEFAULT
+            )
         self.global_query_string_obfuscation_disabled = True  # If empty obfuscation pattern
         self._obfuscation_query_string_pattern = None
         self.http_tag_query_string = True  # Default behaviour of query string tagging in http.url
-        if dd_trace_obfuscation_query_string_pattern != "":
+        if dd_trace_obfuscation_query_string_regexp != "":
             self.global_query_string_obfuscation_disabled = False  # Not empty obfuscation pattern
             try:
                 self._obfuscation_query_string_pattern = re.compile(
```
Comment (majorgreys) [inline ddtrace/settings/config.py:387]:
```suggestion
                self._obfuscation_query_string_regexp = re.compile(
```

**Review by ZStriker19 — APPROVED** [review]

**Pull request merged**